tests/js/client/communication/test-chaos-cluster.js

```diff
@@ -0,0 +1,304 @@
+
+const getDBServers = () => {
+  return getServersByType('dbserver');
+};
+
+const fetchRevisionTree = (serverUrl, shardId) => {
+  let result = request({ method: "POST", url: serverUrl + "/_api/replication/batch", body: {ttl : 3600}, json: true });
+  assertEqual(200, result.statusCode);
+  const batch = JSON.parse(result.body);
+  if (!batch.hasOwnProperty("id")) {
+    throw "Could not create batch!";
+  }
+  
+  result = request({ method: "GET",
+    url: serverUrl + `/_api/replication/revisions/tree?collection=${encodeURIComponent(shardId)}&verification=true&batchId=${batch.id}`});
+  assertEqual(200, result.statusCode);
+  request({ method: "DELETE", url: serverUrl + `/_api/replication/batch${batch.id}`});
+  return JSON.parse(result.body);
+};
+
+const waitForShardsInSync = (cn) => {
+  let start = internal.time();
+  while (true) {
+    if (internal.time() - start > 300) {
+      assertTrue(false, "Shards were not getting in sync in time, giving up!");
+    }
+    let shardDistribution = arango.GET("/_admin/cluster/shardDistribution");
+    assertFalse(shardDistribution.error);
+    assertEqual(200, shardDistribution.code);
+    let collInfo = shardDistribution.results[cn];
+    let shards = Object.keys(collInfo.Plan);
+    let insync = 0;
+    for (let s of shards) {
+      if (collInfo.Plan[s].followers.length === collInfo.Current[s].followers.length) {
+        ++insync;
+      }
+    }
+    if (insync === shards.length) {
+      return;
+    }
+    console.warn("insync=", insync, ", collInfo=", collInfo, internal.time() - start);
+    internal.wait(1);
+  }
+};
+
+const checkCollectionConsistency = (cn) => {
+  const c = db._collection(cn);
+  const servers = getDBServers();
+  const shardInfo = c.shards(true);
+  
+  let failed = false;
+  const getServerUrl = (serverId) => servers.filter((server) => server.id === serverId)[0].url;
+  let tries = 0;
+  do {
+    failed = false;
+    Object.entries(shardInfo).forEach(
+      ([shard, [leader, follower]]) => {
+        const leaderTree = fetchRevisionTree(getServerUrl(leader), shard);
+        leaderTree.computed.nodes = "<reduced>";
+        leaderTree.stored.nodes = "<reduced>";
+        if (!leaderTree.equal) {
+          console.error(`Leader has inconsistent tree for shard ${shard}:`, leaderTree);
+          failed = true;
+        };
+        
+        const followerTree = fetchRevisionTree(getServerUrl(follower), shard);
+        followerTree.computed.nodes = "<reduced>";
+        followerTree.stored.nodes = "<reduced>";
+        if (!followerTree.equal) {
+          console.error(`Follower has inconsistent tree for shard ${shard}:`, followerTree);
+          failed = true;
+        };
+
+        if (!_.isEqual(leaderTree, followerTree)) {
+          console.error(`Leader and follower have different trees for shard ${shard}`);
+          console.error("Leader: ", leaderTree);
+          console.error("Follower: ", followerTree);
+          failed = true;
+        }
```

Comment:
Both `.computed.nodes` and `.stored.nodes` are deleted, so this compares only the metadata if I'm not mistaken. Is this intended, e.g. because we rely on count & hash being different?

Reply:
Yes, this is intentional. We remove the computed and stored nodes since we may want to print the trees, but we don't want to print the 262k buckets! Equality of the trees is checked using the single combined hash and document count.

I've added an according comment.